### DISCUSSION TITLE
the correct way to use source-model watcher

### DISCUSSION THREAD

**Discussion by JimCSuen** [eclipse-glsp/glsp#1460] [Q&A] [answered]
I wish to know the correct way to use source-model watcher. I have added the files in the "client" path like below:
![image](https://github.com/user-attachments/assets/44599d6f-9566-4f4a-a986-f4d494cff18b)

and added related codes in `workflow-diagram-module.ts` like the following:
![image](https://github.com/user-attachments/assets/c1860532-1896-4d96-afe6-430af8f22844)
![image](https://github.com/user-attachments/assets/da05e2bc-9f19-4d9e-a0a0-922645bac028)
![image](https://github.com/user-attachments/assets/5cbea0b5-19f0-4ff3-9fa1-63d3989a9164)

there might be repetition work, but none of them worked. After yarn and yarn start, the running webview does not pop any notifications after i perform any changes directly upon the raw `workflow.wf` file in the `workspace` via notepad, e.g., deleting a node data json block. I don't know if i have done something wrong :(

**Comment by martin-fleck-at** [comment]
Hi @JimCSuen,

The idea of the source model watcher module is that the server sends a `SourceModelChangedAction` when the source model has changed so that the client can inform the user or perform other actions. By default, the module on the client side by default only registers a very [bare bone action handler](https://github.com/eclipse-glsp/glsp-client/blob/61906cf0bcdf68981ead369a34e0112a857f1161/packages/client/src/features/source-model-watcher/source-model-changed-action-handler.ts#L67) that shows that the model has changed as a status and becomes more interesting if you also bind a `ExternalSourceModelChangedHandler` that may be able to do more. We do so, for instance, in the Theia integration where we may also auto-reload the model (see the [TheiaSourceModelChangedHandler](https://github.com/eclipse-glsp/glsp-theia-integration/blob/8ac9d6a8325fd6f67b1aa3f9d7dbf773dcb4c0a0/packages/theia-integration/src/browser/diagram/features/source-model-watcher/theia-source-model-changed-handler.ts#L28)). However, you need to also make sure that someone on the server side is actually sending this action. Currently, only the Java server does so by watching the file, see the [FileWatcher](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/sourcemodelwatcher/FileWatcher.java#L144) implementation. The module itself is already part of the default modules, so if you use the `initializeDiagramContainer` helper function, there is no need to actually add the module again.

I just tested this with the workflow example with the Theia integration and the Java server and I do get the dialog as expected when I change the file on the file system using an external editor:
![image](https://github.com/user-attachments/assets/434e09e1-d18f-4ef2-9ce0-fecf99c3b0c7)

What issues are you facing in particular?

**Reply by JimCSuen** [reply]
Thank you for answering me :) @martin-fleck-at 

For I'm still getting used to this framework, there could be potential faults I have made in binding the handler, which I will keep checking. Anyhow, I'm using the NODE server instead of the JAVA server, is it the source of the problem?

**Reply by JimCSuen** [reply]
Also, it seems that I did not apply any operation at the server side, no wonder nothing happens on the client side. (sorry for my foolishness). So is there any possible bindings for the NODE server?

**Reply by martin-fleck-at** [reply, marked as answer]
No worries, getting used to a new framework is never easy, so I'm happy to answer your questions. 

I just double-checked and I think that is indeed the source of the problem. It seems that the `SourceModelWatcher` API never made it into the Node server, see https://github.com/eclipse-glsp/glsp/issues/1053. You can still implement it yourself [based on the Java implementation](https://github.com/search?q=repo%3Aeclipse-glsp%2Fglsp-server%20SourceModelWatcher&type=code) or by otherwise detecting changes and sending a `SourceModelChangedAction` to the client.

**Reply by JimCSuen** [reply]
Now I understand, I will try the Java server with the source model watcher. Thank you @martin-fleck-at :)
